File: environments/skirmish_crane/paths.py
"""The stable compact codec for tactical movement paths."""

from __future__ import annotations

from .hexes import DIRECTIONS

MAX_PATH_STEPS = 4
# One id per direction sequence of each length, so 6 + 36 + 216 + 1296 = 1554 alongside id 0.
MAX_PATH_ID = sum(len(DIRECTIONS) ** length for length in range(1, MAX_PATH_STEPS + 1))


def encode_path(path: tuple[int, ...] | list[int]) -> int:
    """Encode zero through four direction digits in length then lexical order."""
    try:
        directions = tuple(path)
    except TypeError as error:
        raise ValueError("a path must be an iterable of directions") from error
    if len(directions) > MAX_PATH_STEPS or any(
        type(direction) is not int or direction not in DIRECTIONS for direction in directions
    ):
        raise ValueError("a path must contain zero through four directions numbered 1 through 6")
    if not directions:
        return 0
    return (
        sum(6**power for power in range(1, len(directions)))
        + sum(
            (direction - 1) * 6 ** (len(directions) - offset - 1)
            for offset, direction in enumerate(directions)
        )
        + 1
    )


def decode_path(path_id: int) -> tuple[int, ...]:
    """Decode a stable path id."""
    if isinstance(path_id, bool) or not isinstance(path_id, int) or not 0 <= path_id <= MAX_PATH_ID:
        raise ValueError(f"path id must be an integer from 0 through {MAX_PATH_ID}")
    if path_id == 0:
        return ()
    remaining = path_id - 1
    length = 1
    while remaining >= 6**length:
        remaining -= 6**length
        length += 1
    digits: list[int] = []
    for power in range(length - 1, -1, -1):
        digit, remaining = divmod(remaining, 6**power)
        digits.append(digit + 1)
    return tuple(digits)

File: environments/skirmish_crane/renderer/paths.ts
/**
 * The stable path codec, mirroring the environment's `paths.py`. A path is zero through four
 * direction digits from 1 through 6; ids run in length order and then lexically, with the last
 * direction varying fastest. Id 0 is the empty path, which is the stand-still order.
 *
 * This is part of the student contract, so both sides of it are pinned by tests: the environment's
 * helper pin tests on the Python side, and the mask-agreement suite here.
 */

export const MAX_PATH_STEPS = 4
/** 6 + 6^2 + 6^3 + 6^4 one-through-four-step sequences, alongside id 0. */
export const MAX_PATH_ID = 6 + 6 ** 2 + 6 ** 3 + 6 ** 4

export function encodePath(directions: readonly number[]): number {
  if (directions.length === 0) return 0
  if (directions.length > MAX_PATH_STEPS || directions.some((digit) => digit < 1 || digit > 6)) {
    throw new Error('a Crane Reach path is zero through four directions numbered 1 through 6')
  }
  let id = 1
  for (let length = 1; length < directions.length; length += 1) id += 6 ** length
  for (const [offset, direction] of directions.entries()) {
    id += (direction - 1) * 6 ** (directions.length - offset - 1)
  }
  return id
}

export function decodePath(pathId: number): number[] {
  if (!Number.isInteger(pathId) || pathId < 0 || pathId > MAX_PATH_ID) {
    throw new Error(`a Crane Reach path id is an integer from 0 through ${MAX_PATH_ID}`)
  }
  if (pathId === 0) return []
  let remaining = pathId - 1
  let length = 1
  while (remaining >= 6 ** length) {
    remaining -= 6 ** length
    length += 1
  }
  const directions: number[] = []
  for (let power = length - 1; power >= 0; power -= 1) {
    const place = 6 ** power
    directions.push(Math.floor(remaining / place) + 1)
    remaining %= place
  }
  return directions
}
